sourcegraph/sourcegraph
# Breadcrumbs for single file diff view

Closes #42308.

The main change here is refactoring `RepositoryCompareArea` to be a functional component. Adding in the file path breadcrumbs is only a few additional lines added with one more `useBreadcrumb` call.

There are still a couple of questions I have:

1. I can't actually seem to get hovering features to work - or they at least don't seem to be working the same for me as they are on production? I referenced [this other functional component](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@lrh/single-file-diff-view-breadcrumbs/-/blob/client/web/src/enterprise/batches/close/BatchChangeCloseChangesetsList.tsx?L95-133) for how to set it up.
1. I wasn't sure how to replicate the error handling that originally came from the `state` object - see [this TODO](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@lrh/single-file-diff-view-breadcrumbs/-/blob/client/web/src/repo/compare/RepositoryCompareArea.tsx?L163-168). That `error` property was defined as part of the extended interface for the components state, and I wasn't finding where it may have been getting set.
2. The "Copy path to clipboard" button is rendered by [CopyPathAction](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@lrh/single-file-diff-view-breadcrumbs/-/blob/client/web/src/repo/actions/CopyPathAction.tsx), which uses `parseBrowserRepoURL` to determine what the "path" to copy should be. We could update `parseBrowserRepoURL` to also look for the `pathName` search param - is that approach okay? Or is that starting to spread these changes out too far?

## Test plan

1. Check that there is no regression to compare page after refactor. ([Local dev link](https://sourcegraph.test:3443/github.com/sourcegraph/sourcegraph/-/compare/...main-dry-run/mv/debug-integration-test))
4. Confirm that the file path breadcrumbs are rendered when viewing a single file. ([Local dev link](https://sourcegraph.test:3443/github.com/sourcegraph/sourcegraph/-/compare/...main-dry-run/mv/debug-integration-test?filePath=enterprise%2Fcmd%2Ffrontend%2Finternal%2Fauthz%2Fresolvers%2Fresolver.go&visible=1))



## App preview:

- [Web](https://sg-web-lrh-single-file-diff-view.onrender.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
